### ISSUE TITLE
[expression-language] unclear return type on evaluate

### ISSUE THREAD

**Issue by williamdes** [symfony/symfony#39134]
I had to make a patch for an unexpected false returned.

See: https://github.com/phpmyadmin/motranslator/pull/38

https://github.com/symfony/expression-language/blob/6fca724c3616e9fb49a06f183e33d886677fa0d1/ExpressionLanguage.php#L63

**Description**  
Unclear method return

Scrutinizer is also puzzled: https://scrutinizer-ci.com/g/phpmyadmin/motranslator/issues/master/files/src/Translator.php?orderField=path&order=asc&honorSelectedPaths=0

**How to reproduce**  
https://github.com/phpmyadmin/motranslator/blob/16712bfec27a53217721b3d51983e541d4a3eba4/src/Translator.php#L320

**Possible Solution**  
Update the phpdoc

**Additional context**  
See the diff of https://github.com/phpmyadmin/motranslator/pull/38 to have a test case

**Comment by wouterj** [comment]
Hi! 

Can you be a little more clear on what is wrong in the PHPdoc? The `ExpressionLanguage::evaluate()` method calls the `evaluate()` method of the node classes. No method has a hardcoded `false`. So as far as I can see `false` is only returned when the expression evaluated to `false`.
In that case, the PHPdoc is correct and your fix was the correct one: motranslator apparently has an expression that can evaluate false, but it didn't handle `false` correctly.

**Comment by williamdes** [comment]
> Hi!
> 
> Can you be a little more clear on what is wrong in the PHPdoc? The `ExpressionLanguage::evaluate()` method calls the `evaluate()` method of the node classes. No method has a hardcoded `false`. So as far as I can see `false` is only returned when the expression evaluated to `false`.
> In that case, the PHPdoc is correct and your fix was the correct one: motranslator apparently has an expression that can evaluate false, but it didn't handle `false` correctly.

Hi !

Nothing is wrong, it is just unclear what the method returns.
` * @return mixed The result of the evaluation of the expression`
Neither phpstan or any tool I know is able to find out correctly what the mixed is really.

So I would like that it is improved to be explicit and say what types can be returned so people can implement correctly the securities around your code to avoid "motranslator apparently has an expression that can evaluate false, but it didn't handle false correctly."
We could not handle something that was not expected ;)

I hope you did understand me

**Comment by derrabus** [comment]
> ` * @return mixed The result of the evaluation of the expression`

This is as precise as it gets. The return type depends on the actual expression that is evaluated.

**Comment by williamdes** [comment]
> > ` * @return mixed The result of the evaluation of the expression`
> 
> This is as precise as it gets. The return type depends on the actual expression that is evaluated.

Well you could at least say `* @return mixed|false The result of the evaluation of the expression or false on failure`

**Comment by derrabus** [comment]
> Well you could at least say `* @return mixed|false The result of the evaluation of the expression or false on failure`

`mixed` does include `false` if I'm not mistaken. Also, that description would be inaccurate as @wouterj already noted:

> as far as I can see `false` is only returned when the expression evaluated to `false`.

**Comment by williamdes** [comment]
> mixed does include false if I'm not mistaken. Also, that description would be inaccurate as @wouterj already noted:

Yup, I it is true but precise documentation is for humans and static analysers


I think you sould have a look to my example and let me know if there is a bug or not on your side on why it did return false.

**Comment by wouterj** [comment]
The point here is not whether or not `mixed` includes false. The point is that there is *no* hardcoded return type or value in any situation in the component as far as I can see. There is no "false on failure". There might be one in the specific case of motranslator (because a method/function defined by motranslator used in the expression might return false on failure), but there is no such thing in the component itself afaics.

**Comment by williamdes** [comment]
Okay, I will have to dig why you did return me a `false` ~~tomorrow~~.
Edit: :hourglass: delayed

**Comment by wouterj** [comment]
Thank you! The code you linked to is a bit hard to follow without cloning and running it. Please let us know if you can't find it and I'll try to run your example.

**Comment by stof** [comment]
your own test case has `Plural-Forms: nplurals=2; plural=n != 1;` as the header from which you return an expression, and `n != 1` is indeed evaluating to a boolean. If you expect a string as the evaluated result, the issue is in the fact that you accept such expression in your tool, not on the ExpressionLanguage logic.
Also, if I write `plural=null` in there, suddently you'll get `null` and that's not an integer either. That's not only about `false` or anything else. And that's precisely why the component is typed as `return mixed`: it can return *anything*.

Btw, the reason why Scrutinizer gets confused is because it decided (for unknown reason) that `ExpressionLanguage::evaluate` returns `array` and not `mixed` (and of course an array cannot be equal to `false`). That looks like a bug in Scrutinizer (maybe they try to analyze the `ExpressionLanguage` source code instead of trusting the phpdoc and fail at understanding it properly).

I'm closing this as there is no issue in Symfony. The existing phpdoc is the most accurate one.

**Comment by williamdes** [comment]
Hi,
Thank you for all the feedback, I changed my patch to cast to an int the result in https://github.com/phpmyadmin/motranslator/commit/7a8e6a21fc47f89b74515e4c475f83f1bd97f3a6.
And I added I think almost all data cases we can have to test phpunit test case

**Comment by jderusse** [comment]
I commented the commit. I think your issue is in your test suite the expression returns a bool, while you method expect an integer.

`Plural-Forms: nplurals=2; plural=n != 1;'` <= wrong `n != 1` returns a bool
`Plural-Forms: nplurals=3; plural=n%10==1 && n%100!=11 ? 0 : n` <= OK returns `0` or `N`

I don't know if expression in MO files should return a integer or a boolean, and in case of boolean, does `true` means `0`, `1` or `N`?, but IMHO you should have a specific code to handle such situation.

**Comment by williamdes** [comment]
> I commented the commit. I think your issue is in your test suite the expression returns a bool, while you method expect an integer.
> 
> `Plural-Forms: nplurals=2; plural=n != 1;'` <= wrong `n != 1` returns a bool
> `Plural-Forms: nplurals=3; plural=n%10==1 && n%100!=11 ? 0 : n` <= OK returns `0` or `N`
> 
> I don't know if expression in MO files should return a integer or a boolean, and in case of boolean, does `true` means `0`, `1` or `N`?, but IMHO you should have a specific code to handle such situation.

Hi
Thank you for the feedback on our implementation
@mauriciofauth do you have thoughts on this?
